anonlink/serialization.py

```diff
@@ -0,0 +1,221 @@
+# All integers are unsigned. All values are *little-endian* (yes, this
+# breaks convention). The number of candidate matchings is not included
+# on purpose: if needed, it can be computed from the length of the file.
+
+
+def _entry_struct(
+    sim_t_size: int,
+    dset_i_t_size: int,
+    rec_i_t_size: int
+) -> _struct.Struct:
+    try:
+        sim_t = _STRUCT_FLOAT_LEN_TO_FMT[sim_t_size]
+    except KeyError:
+        msg = f'floats of {sim_t_size} bytes are not supported'
+        raise ValueError(msg) from None
+    try:
+        dset_i_t = _STRUCT_UINT_LEN_TO_FMT[dset_i_t_size]
+    except KeyError:
+        msg = f'indices of {dset_i_t_size} bytes are not supported'
+        raise ValueError(msg) from None
+    try:
+        rec_i_t = _STRUCT_UINT_LEN_TO_FMT[rec_i_t_size]
+    except KeyError:
+        msg = f'indices of {rec_i_t_size} bytes are not supported'
+        raise ValueError(msg) from None
+    return _struct.Struct(f"<{sim_t}2{dset_i_t}2{rec_i_t}")
+    
+
+def _dump_from_iterable(
+    f: _typing.BinaryIO,
+    sim_t_size: int,
+    dset_i_t_size: int,
+    rec_i_t_size: int,
+    iterable: _CandidatePairIter
+) -> None:
+    # Fail without writing if the parameters are incorrect.
+    entry_struct = _entry_struct(
+        sim_t_size, dset_i_t_size, rec_i_t_size)
+    
+    # Write header.
+    f.write(_HEADER_STRUCT.pack(1, sim_t_size, dset_i_t_size, rec_i_t_size))
+    
+    for entry in iterable:
+        f.write(entry_struct.pack(*entry))
+
+
+def _entries_iterable(
+    f: _typing.BinaryIO,
+    entry_struct: _struct.Struct
+) -> _CandidatePairIter:
+    buffer = f.read(entry_struct.size)
+    while buffer:
+        if len(buffer) != entry_struct.size:
+            raise ValueError('ran out of input')
+        yield _typing.cast(_CandidatePair, entry_struct.unpack(buffer))
+        buffer = f.read(entry_struct.size)
+
+
+def _load_to_iterable(
+    f: _typing.BinaryIO
+) -> _typing.Tuple[_CandidatePairIter, int, int, int]:
+    # Make sure that f.read(n) returns exactly n bytes.
+    if isinstance(f, _io.RawIOBase):
+        f = _io.BufferedWriter(f)
+    elif not isinstance(f, _io.BufferedIOBase):
+        msg = 'stream must be instance of BufferedIOBase or RawIOBase'
+        raise ValueError(msg)
+
+    header_bytes = f.read(_HEADER_STRUCT.size)
+    if len(header_bytes) != _HEADER_STRUCT.size:
+        raise ValueError('ran out of input')
+    version, sim_t_size, dset_i_t_size, rec_i_t_size = (
+        _HEADER_STRUCT.unpack(header_bytes))
+    if version != 1:
+        raise ValueError('unsupported version of serialized file')
+
+    # This may throw if the specified format isn't supported on this
+    # platform.
+    entry_struct = _entry_struct(sim_t_size, dset_i_t_size, rec_i_t_size)
+
+    return (_entries_iterable(f, entry_struct),
+            sim_t_size, dset_i_t_size, rec_i_t_size)
+
+
+def dump_candidate_pairs(
+    f: _typing.BinaryIO,
+    candidate_pairs: _typechecking.CandidatePairs
+) -> None:
```

Comment:
Nitpicky but I think we should be consistent with the Python standard library where the object to dump precedes the file pointer: e.g. [ `json.dump(obj, fp)`](https://docs.python.org/3/library/json.html#json.dump)

Reply:
Good point!